Stackdriver: Rename Stackdriver to Google Cloud Monitoring

papagian opened the pull request:
<!--

Thank you for sending a pull request! Here are some tips:

1. If this is your first time, please read our contribution guide at https://github.com/grafana/grafana/blob/master/CONTRIBUTING.md

2. Ensure you include and run the appropriate tests as part of your Pull Request.

3. In a new feature or configuration option, an update to the documentation is necessary. Everything related to the documentation is under the docs folder in the root of the repository.

4. If the Pull Request is a work in progress, make use of GitHub's "Draft PR" feature and mark it as such.

5. If you can not merge your Pull Request due to a merge conflict, Rebase it. This gets it in sync with the master branch.

6. Name your PR as "<FeatureArea>: Describe your change", e.g. Alerting: Prevent race condition. If it's a fix or feature relevant for the changelog describe the user impact in the title. The PR title is used to auto-generate the changelog for issues marked with the "add to changelog" label.

-->

**What this PR does / why we need it**:
It renames `Stackdriver` datasource to `Google Cloud Monitoring`

It adds legacy support for existing:
- datasources
- provisioned datasources
- queries (metric/SLO)
- `stackdriver auto` alignment period (introduces also `cloud monitoring auto` alignment period)
- alerts
- annotations
- template variables

**Which issue(s) this PR fixes**:

<!--

* Automatically closes linked issue when the Pull Request is merged.

Usage: "Fixes #<issue number>", or "Fixes (paste link of issue)"

-->

Fixes #25668

**Special notes for your reviewer**:
Existing dashboards should work without any change.
Imported dashboards (exported from previous grafana versions) should load successfully.

A complementary PR for renaming static docs images will follow up.

papagian committed:
Update backend

papagian committed:
Update frontend

papagian committed:
Legacy support existing dashboards

papagian committed:
More lecacy fixes

papagian committed:
Update docs

papagian committed:
Fix serving cloud monitoring logo

papagian committed:
Merge remote-tracking branch 'origin/master' into rename-stackdriver

oddlittlebird commented:
Docs look good.

daniellee commented:
@papagian Great stuff. Will review tomorrow.

marefr commented on pkg/models/datasource.go:
```diff
@@ -9,22 +9,22 @@ import (
+	DS_MSSQL            = "mssql"
+	DS_ACCESS_DIRECT    = "direct"
+	DS_ACCESS_PROXY     = "proxy"
+	DS_CLOUD_MONITORING = "cloud-moniroting"
```
```suggestion
	DS_CLOUD_MONITORING = "cloud-monitoring"
```

marefr commented on pkg/tsdb/query_endpoint.go:
```diff
@@ -20,6 +20,10 @@ func init() {
 }
 
 func getTsdbQueryEndpointFor(dsInfo *models.DataSource) (TsdbQueryEndpoint, error) {
+	if dsInfo.Type == "stackdriver" { // legacy
```
Having this logic here is a code smell. Instead of this, can we register query endpoint for both legacy stackdriver and new cloud monitoring? See my comment of plugin.json

marefr commented on public/app/features/datasources/state/actions.ts:
```diff
@@ -62,7 +62,13 @@ export const initDataSourceSettings = (
-      const dataSourceMeta = dependencies.getDataSourceMeta(getState().dataSources, dataSource.type);
+
+      let dsType = dataSource.type;
+      if (dsType === 'stackdriver') {
```
Having this logic here feels like a code smell. Can we handle the legacy aspect from backend solely?

marefr commented on public/app/plugins/datasource/cloud-monitoring/components/AlignmentPeriods.tsx:
```diff
@@ -31,6 +31,11 @@ export const AlignmentPeriods: FC<Props> = ({
     label: ap.text,
   }));
 
+  if (alignmentPeriod === 'stackdriver-auto') {
```
Can we handle this on the backend instead?

papagian replied:
It is [handled](https://github.com/grafana/grafana/blob/3037bda44908e572411d7b1447f460307837a47e/pkg/tsdb/cloudmonitoring/cloudmonitoring.go#L309) in the backend. This is required here because the `stackdriver-auto` value is no more among the available alignment periods so its value [appears empty](https://github.com/grafana/grafana/blob/3037bda44908e572411d7b1447f460307837a47e/public/app/plugins/datasource/cloud-monitoring/components/AlignmentPeriods.tsx#L45) in the query editor.
An alternative would be to keep `stackdriver-auto` as hidden alignment period. I will try that.

marefr commented on public/app/plugins/datasource/cloud-monitoring/plugin.json:
```diff
@@ -1,7 +1,7 @@
+  "name": "Cloud Monitoring",
   "type": "datasource",
-  "id": "stackdriver",
+  "id": "cloud-monitoring",
```
It generates a lot of extra checks and handling changing the id of the plugin. Thinking about introducing a new property in plugin model json called `legacyId`or something that can hold "stackdriver" and be used in plugins package. Thoughts?

marefr commented on public/app/plugins/datasource/cloud-monitoring/plugin.json:
```diff
@@ -1,7 +1,7 @@
 {
-  "name": "Stackdriver",
+  "name": "Cloud Monitoring",
```
I've read Cloud Monitoring in several places and in some places Google Cloud Monitoring. Think we should try to be consistent.

marefr commented on pkg/api/datasources.go:
```diff
@@ -316,6 +317,10 @@ func (hs *HTTPServer) CallDatasourceResource(c *models.ReqContext) {
 }
 
 func convertModelToDtos(ds *models.DataSource) dtos.DataSource {
+	if ds.Type == "stackdriver" { // legacy
```
Code smell, this logic should reside in plugins package. See my comment on plugin.json

marefr reviewed (changes requested):
Great. Only reviewed the code so far. Really tricky to change the plugin id and be backward and forward compatible - I added some comments about this. Feels like we're lacking a concept in plugins that allow a plugin id to be renamed. Added some suggestions as well.

Was thinking of database migration, but think there's still a problem with alert rules having the data source type embedded. Have you considered that?

Would like to get some more eyes on this and whether changing the plugin id is absolutely necessary.

papagian commented:
> Great. Only reviewed the code so far. Really tricky to change the plugin id and be backward and forward compatible - I added some comments about this. Feels like we're lacking a concept in plugins that allow a plugin id to be renamed. Added some suggestions as well.
> 
> Was thinking of database migration, but think there's still a problem with alert rules having the data source type embedded. Have you considered that?
> 
> Would like to get some more eyes on this and whether changing the plugin id is absolutely necessary.

I share more or less the same concerns. I had also started working in a separate branch and experimenting not to renaming the plugin id but there were still things that had to take care of so I chose to go forward with the present solution. However I will put some effort and investigate the alternative in order to evaluate both approaches and pick the cleanest and safest.

papagian commented on docs/sources/features/datasources/cloudmonitoring.md:
```diff
@@ -50,45 +50,45 @@ The following APIs need to be enabled first:
 1. First of all, you need to create a Service Account that can be used by the GCE virtual machine. See detailed instructions on how to do that [here](https://cloud.google.com/compute/docs/access/create-enable-service-accounts-for-instances#createanewserviceaccount).
 2. Make sure the GCE virtual machine instance is being run as the service account that you just created. See instructions [here](https://cloud.google.com/compute/docs/access/create-enable-service-accounts-for-instances#using).
-3. Allow access to the `Stackdriver Monitoring API` scope. See instructions [here](changeserviceaccountandscopes).
+3. Allow access to the `Cloud Monitoring Monitoring API` scope. See instructions [here](changeserviceaccountandscopes).
```
I have no idea. I have browsed the file history but I was unable to figure out where it originally pointed. Therefore I was thinking of removing it completely.

papagian committed:
Apply suggestions from code review

Co-authored-by: Marcus Efraimsson <marcus.efraimsson@gmail.com>

papagian committed:
Place doc images to a new directory

papagian commented:
Closing this in favour of #25807

Present comment are taken into consideration in the new PR.
